pkg/reconciler/v1alpha1/taskrun/validate_test.go

```diff
@@ -0,0 +1,267 @@
+						Value: "somethingfun",
+					},
+				},
+			},
+		},
+	}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-taskrun-bad-resourcetype",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.TaskRunSpec{
+			TaskRef: v1alpha1.TaskRef{
+				//
+				Name: "unit-task-bad-resourcetype",
+			},
+			Inputs: v1alpha1.TaskRunInputs{
+				Resources: []v1alpha1.TaskRunResourceVersion{
+					v1alpha1.TaskRunResourceVersion{
+						Name: "testimageinput",
+						ResourceRef: v1alpha1.PipelineResourceRef{
+							Name: "git-test-resource",
+						},
+					},
+				},
+			},
+		},
+	}}
+
+	ts := []*v1alpha1.Task{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "unit-test-task",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.TaskSpec{
+			Inputs: &v1alpha1.Inputs{
+				Resources: []v1alpha1.TaskResource{{}},
+			},
+		},
+	}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "unit-task-wrong-input",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.TaskSpec{
+			Inputs: &v1alpha1.Inputs{
+				Resources: []v1alpha1.TaskResource{{
+					Name: "testinput",
+				}},
+			},
+		},
+	}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "unit-task-wrong-output",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.TaskSpec{
+			Outputs: &v1alpha1.Outputs{
+				Resources: []v1alpha1.TaskResource{{
+					Name: "testoutput",
+				}},
+			},
+		},
+	}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "unit-task-multiple-params",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.TaskSpec{
+			Inputs: &v1alpha1.Inputs{
+				Params: []v1alpha1.TaskParam{{
+					Name: "foo",
+				}, {
+					Name: "bar",
+				}},
+			},
+		},
+	}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "unit-task-bad-resourcetype",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.TaskSpec{
+			Inputs: &v1alpha1.Inputs{
+				Resources: []v1alpha1.TaskResource{{
+					Name: "testimageinput",
+					Type: v1alpha1.PipelineResourceTypeImage,
+				}},
+			},
+		},
+	}}
+
+	rr := []*v1alpha1.PipelineResource{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "git-test-resource",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.PipelineResourceSpec{
+			Type: v1alpha1.PipelineResourceTypeGit,
+			Params: []v1alpha1.Param{{
+				Name:  "foo",
+				Value: "bar",
+			}},
+		},
+	}}
+
+	tcs := []struct {
+		name    string
+		taskrun *v1alpha1.TaskRun
+		reason  string
+	}{
+		{
+			name:    "bad-input-source-bindings",
+			taskrun: trs[0],
+			reason:  "input-source-binding-to-invalid-resource",
+		}, {
+			name:    "bad-output-source-bindings",
+			taskrun: trs[1],
+			reason:  "output-source-binding-to-invalid-resource",
+		}, {
+			name:    "bad-inputkey",
+			taskrun: trs[2],
+			reason:  "bad-input-mapping",
+		}, {
+			name:    "bad-outputkey",
+			taskrun: trs[3],
+			reason:  "bad-output-mapping",
+		}, {
+			name:    "param-mismatch",
+			taskrun: trs[2],
+			reason:  "input-param-mismatch",
+		}, {
+			name:    "resource-mismatch",
+			taskrun: trs[3],
+			reason:  "input-resource-mismatch",
```

Comment:
this reason is not used anywhere. can we drop it?

Reply:
we can drop it, i think it supposed to be a helpful descriptor for the test.  I am copying the test format used in PipelineRun.  I will leave this as is for now.